### ISSUE TITLE
Proposed OVAL Updates For Microsoft Internet Information Server (IIS) Settings Using AppCmd.exe - appcmdlistconfig Test

### ISSUE THREAD

**Issue by bryanlwilson** [OVAL-Community/OVAL#77]
**Abstract**
OVAL 5.11.2 lacks a robust, straight-forward, and efficient mechanism to query Microsoft Internet
Information Server (versions 7.0 – 10) configuration settings. The existing OVAL cmdlet test is
inadequate as it fails to provide adequate data resolution. For instance, if an Internet Information
Server (IIS) installation hosts multiple websites, the cmdlet test results do not enable the user to
determine from which website a reported setting was collected.
The Microsoft AppCmd.exe application is the primary method for querying and managing IIS
servers. We propose two new OVAL tests, the appcmd and appcmdlistconfig tests to query IIS
data using AppCmd.exe.
This ticket defines the proposed appcmdlistconfig test. See a related ticket (proposal) covering the appcmd test.

**Link to Proposal**
The proposal is attached to this ticket as a PDF file. It is dated May 03, 2019, and is updated to version 1.1. Please review document for OVAL Update.

**Additional context**

[win-def_appcmdlistconfig_addition_v1.1.pdf](https://github.com/OVAL-Community/OVAL/files/3142366/win-def_appcmdlistconfig_addition_v1.1.pdf)
[win-def_appcmdlistconfig_addition_v1.1.zip](https://github.com/OVAL-Community/OVAL/files/3142395/win-def_appcmdlistconfig_addition_v1.1.zip)

**Comment by wmunyan** [comment]
Can you please add the declarations of the `EntityObjectAppCmdObjectType` and `EntityItemAppCmdObjectType` elements?  Are these enumerations?

**Comment by wmunyan** [comment]
Could you please add the declarations of the `Entity(Object|State|Item)AppCmdListConfigIdentifierType`?

**Comment by bryanlwilson** [comment]
Good catch.  That should of been included in the appcmdlistconfig_updates_windows-system-characteristics-schema.xsd file contained in "schema changes" folder in zip file.

I will update the package and resubmit.

**Comment by wmunyan** [comment]
Quick comments:
The `identifier` element in the `appcmdlistconfig_object` is listed as `minOccurs=0`, but I feel like convention would be to make all object elements have `minOccurs=1`, but also set `nillable=true` in order to effectively remove it.  The documentation for that element also references the `xsi:nil` attribute already.

Second, I feel like the `result_descriptor` element may be redundant?  Wouldn't the `identifier` element (at least in the system characteristics) identify the information/location of the found result?  For example, if an object's `identifier` was a pattern match on, say, `[Dd]efault.*`, wouldn't each resulting `appcmdlistconfig_item`s have an `identifier` for each matching value (such as "Default Web Site", "Default Bill Site", etc)?

CIS did a more elaborate extension schema for IIS a while back, so we'll have to check how we can map that to the new schema.  More to follow on that one.

Cheers, 
-Bill M.

**Comment by bryanlwilson** [comment]
sounds like i may look into more updates... for speed i am including the first request as a comment. Next comment to come.....

**Comment by bryanlwilson** [comment]
<xsd:complexType name="EntityItemAppCmdListConfigIdentifierType">
          <xsd:annotation>
               <xsd:documentation>Restricts a string value to a set of allowed appcmd objects.</xsd:documentation>
          </xsd:annotation>
          <xsd:simpleContent>
               <xsd:restriction base="oval-sc:EntityItemStringType">
					<xsd:enumeration value="Site">
						  <xsd:annotation>
								<xsd:documentation>Administration of virtual sites</xsd:documentation>
						  </xsd:annotation>
					</xsd:enumeration>
					<xsd:enumeration value="VDir">
						  <xsd:annotation>
								<xsd:documentation>Administration of virtual directories</xsd:documentation>
						  </xsd:annotation>
					</xsd:enumeration>
					<xsd:enumeration value="Apppool">
						  <xsd:annotation>
								<xsd:documentation>Administration of application pools</xsd:documentation>
						  </xsd:annotation>
					</xsd:enumeration>
					<xsd:enumeration value="Webserver">
						  <xsd:annotation>
								<xsd:documentation>Server level configuration</xsd:documentation>
						  </xsd:annotation>
					</xsd:enumeration>
					<xsd:enumeration value="">
						<xsd:annotation>
							  <xsd:documentation>The empty string value is permitted here to allow for empty elements associated with variable references.</xsd:documentation>
						</xsd:annotation>
					</xsd:enumeration>
               </xsd:restriction>
          </xsd:simpleContent>
     </xsd:complexType>

**Comment by bryanlwilson** [comment]
I have updated the appcmdlistconfig_updates_windows-system-characteristics-schema.xsd to add the EntityItemAppCmdListConfigIdentifierType.  Please see zip file attached. I will update the complete package soon, want to make sure there aren't any other needed updates.

[appcmdlistconfig_updates_windows_schema.zip](https://github.com/OVAL-Community/OVAL/files/3158634/appcmdlistconfig_updates_windows_schema.zip)

**Comment by bryanlwilson** [comment]
responding to:
Second, I feel like the result_descriptor element may be redundant? Wouldn't the identifier element (at least in the system characteristics) identify the information/location of the found result? For example, if an object's identifier was a pattern match on, say, [Dd]efault.*, wouldn't each resulting appcmdlistconfig_items have an identifier for each matching value (such as "Default Web Site", "Default Bill Site", etc)?

BW response.
The result_descriptor is needed because when appcmd is used to gather data the linkage between the location and the settings must be maintained for usable results.  In your example you would need both the result_value and result_descriptor in order to correctly capture what the actual settings is on "Default Web Site" and what actual setting is on "Default Bill Site".   In this example the 2 collected items show that the  "Default Web Site" has a settings of 'Batch" and and the "Default Bill Site" has a setting of 'ClearText'.

Here is an example:
Collected Item Properties:
- identifier_type equals 'Site'
- identifier equals '.*'
- section equals 'system.webServer/security/authentication/anonymousAuthentication'
- parameter equals 'logonMethod'
- result_descriptor equals 'Default Web Site'
- result_value equals 'Batch'

Collected Item Properties:
- identifier_type equals 'Site'
- identifier equals '.*'
- section equals 'system.webServer/security/authentication/anonymousAuthentication'
- parameter equals 'logonMethod'
- result_descriptor equals 'Default Bill Site'
- result_value equals 'ClearText'

**Comment by bryanlwilson** [comment]
responding to:
Second, I feel like the result_descriptor element may be redundant? Wouldn't the identifier element (at least in the system characteristics) identify the information/location of the found result? For example, if an object's identifier was a pattern match on, say, [Dd]efault.*, wouldn't each resulting appcmdlistconfig_items have an identifier for each matching value (such as "Default Web Site", "Default Bill Site", etc)?

BW response.
I concur with your requested update. In the original design result_descriptor is needed because when appcmd is used to gather data the linkage between the location and the settings must be maintained for usable results. In your example you would need both the result_value and result_descriptor in order to correctly capture what the actual settings is on "Default Web Site" and what actual setting is on "Default Bill Site". In this example the 2 collected items show that the "Default Web Site" has a settings of 'Batch" and and the "Default Bill Site" has a setting of 'ClearText'.

To document how collected items would now look see the "current" item design and "request update" item design
"current" item design with 2 sites.
Collected Item Properties:
    identifier_type equals 'Site'
    identifier equals '.*'
    section equals 'system.webServer/security/authentication/anonymousAuthentication'
    parameter equals 'logonMethod'
    result_descriptor equals 'Default Web Site'
    result_value equals 'Batch'
Collected Item Properties:
    identifier_type equals 'Site'
    identifier equals 'Default Web Site'
    section equals 'system.webServer/security/authentication/anonymousAuthentication'
    parameter equals 'logonMethod'
    result_descriptor equals 'Default Bill Site'
    result_value equals 'ClearText'

"request update" item design with same 2 sites
Updates:
1- identifier is populated with previous "result_descriptor" data
2 - "result_value" renamed to "result"
3 - "result_descriptor" has been removed.

Collected Item Properties:
    identifier_type equals 'Site'
    identifier equals 'Default Web Site'
    section equals 'system.webServer/security/authentication/anonymousAuthentication'
    parameter equals 'logonMethod'
    result equals 'Batch'
Collected Item Properties:
    identifier_type equals 'Site'
    identifier equals 'Default Bill Site'
    section equals 'system.webServer/security/authentication/anonymousAuthentication'
    parameter equals 'logonMethod'
    result equals 'ClearText'

**Comment by bryanlwilson** [comment]
Responding to
The identifier element in the appcmdlistconfig_object is listed as minOccurs=0, but I feel like convention would be to make all object elements have minOccurs=1, but also set nillable=true in order to effectively remove it. The documentation for that element also references the xsi:nil attribute already

BW Response
I concur with this update as well.   I will now update my schema files and resubmit for review. Once agreed the updates seem satisfactory i will update and resubmit the complete package.

Finally,
Thank you very much for this feedback.  Glad things seem to be moving in the right direction for OVAL updates.  This type of feedback is great.  I am going to concentrate on updating appcmdlistconfig first before appcmd.

**Comment by bryanlwilson** [comment]
I have updated the appcmdlistconfig schema.  Please review and let me know if the updates have been applied correctly.

[appcmdlistconfig_updates_windows-schema.zip](https://github.com/OVAL-Community/OVAL/files/3158808/appcmdlistconfig_updates_windows-schema.zip)

**Comment by wmunyan** [comment]
I'll preface by saying that I _think_ I'm looking at the latest version of the schema.  It looks like the `nillable` attribute is still missing from any elements that may require it.  Also, the element descriptions still reference a "relative dn", which seems like its a cut/paste from some other elements.

**Comment by solind** [comment]
Hi @wmunyan and @bryanlwilson,

I see that appcmd.exe has an ```/xml``` command-line option.  Naturally I am predisposed to having an XPath-compatible test for anything that can be expressed as XML.

This has the benefit of being somewhat simpler in that (for instance) there would be no need to specify any identifier, let alone wonder whether it should be made nilable.

What do you think of this idea?

**Comment by bryanlwilson** [comment]
https://github.com/OVAL-Community/OVAL/issues/76#issuecomment-512399327

**Comment by Memnarch** [comment]
just took a quick look, and i'd suggest to not put the implementationdetail "appcmd.exe" into the type or somewherevisible.
In addition to appcmd.exe, Microsoft provides a com api. Using IAppHostAdminManager it seems you can accomplish the same tasks. In my opinion it's up to the engine implementor how the IIS7+ information are served, as loong as it meets the criteria.

https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/iapphostadminmanager-interface

I propose to change the name to something that just has IIS, in some way, in it's name.

**Comment by solind** [comment]
@Memnarch That is an excellent idea; there is also the Microsoft.Web.Administration.ServerManager C# class.  See: https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.administration.servermanager?view=iis-dotnet

Working with command-line programs like AppCmd.exe can be problematic.  Bryan and I explored the XML option off-list, but that turned out to preclude access to default values.  So, aligning with a programmatic means of accessing this data would be far superior.

**Comment by bryanlwilson** [comment]
Response to IIS / Appcmd.exe related comments regarding utilizing Microsoft.Web.Administration.dll instead of appcmd.exe calls to gather IIS configuration settings.

I have researched attempting to utilize Microsoft.Web.Administration.dll to gather already implemented appcmd and appcmdlist config objects for already created DISA content to attempt to utilize the new .dll solution.  This resulted in roughly 20% of the 50+ definitions that utilize appcmd.exe related tests to were lost in this translation and another 10 objects (Issue 4 in readme file included) have returned data that is accurate but doesn't contain setting masks that appcmd.exe returns, which sometimes makes content a little harder to write and not as readable.  In updated IIS Site and Server content I have included links to the setting masks where Microsoft documentation was found (included in updated content file object comments).

Not to get too far into weeds but after going through this process of converting data gathering it was pretty apparent that the format of the dll calls were different based on whether it was an existing appcmd vs appcmdlistconfig related object. By searching the updated disa content site and server xml content documents I am about to share on GIT hub, you can see all the appcmd.exe related objects show the corresponding appcmd.exe command and now include the call to Microsoft.Web.Administration.dll. Futhermore I have included in the already referenced readme file attached to this email notes about the appcmd use vs Microsoft.Web.Administration.dll use and how the calls are made.

Testing Totals
of the 16 Servers object - 4 appcmd.exe related object cannot be translated to easily utilize (or functionally utilize) Microsoft.Web.Administration.dll
of the 40 Site objects - 7 appcmd.exe related objects cannot be translated to easily utilize (or functionally utilize) Microsoft.Web.Administration.dll
of the 45 objects that I could utilize, 10 checks work however the data mask doesn’t appear to exist, for the most part I marked the data mask issue as ok (success, check can be implemented with .dll) with the one exception noted as ISSUE 4 in attached readme.
Note: I have attached a readme to this email with further details of the issues I saw with converting current DISA STIG appcmd.exe content.  Also I plan on updating gethub ticket to include my updated content with comments.

So where do we go from here?
Option 1 – Someone else picks up what I have done and try and prove all the checks can be captured with Microsoft.Web.Administration.dll.
Option 2 – Approve proposals for appcmd and appcmdlistconfig after minor suggested updates but overall design stays same.  
Option 3 – Abstract the appcmd and appcmdlistconfig tests to be one IIS test.  Scanner tools can utilize this updated test and object to build IIS data gathering tests.  This may make the job harder for content author and may result in different scap approved tools reporting slightly different format of settings depending on what internal data gathering process is used.



[AppcmdvsWebAdminDLL_Readme.txt](https://github.com/OVAL-Community/OVAL/files/3478523/AppcmdvsWebAdminDLL_Readme.txt)
[U_IIS_8-5_STIG_V1R1_Benchmark-oval_updated20190807.zip](https://github.com/OVAL-Community/OVAL/files/3478525/U_IIS_8-5_STIG_V1R1_Benchmark-oval_updated20190807.zip)

**Comment by solind** [comment]
Hi Bryan,

I have fully examined ISSUE 1 and made the following observations:

The /text:all switch in this context merely dumps the XML element tree to a text format.  So, you can simply explore the elements and attributes to find whatever it is that you’re looking for.  To obtain the custom fields for oval:mil.spawar.microsoft.iis85.site:obj:100, for example, use:
$iis.Sites["Default Web Site"].GetChildElement("logFile").GetChildElement("customFields")

To obtain all the fileExtension attribute values for oval:mil.spawar.microsoft.iis85.site:obj:130 or oval:mil.spawar.microsoft.iis85.server:obj:170, use:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.webServer/staticContent").GetCollection() | %{$_.Attributes["fileExtension"]} | %{$_.Value}

Note that I used the GetCollection() method of the ConfigurationSection class as advised by the Microsoft documentation here (NB the Remarks section):
https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.administration.configurationelement.getcollection?view=iis-dotnet#Microsoft_Web_Administration_ConfigurationElement_GetCollection <https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.administration.configurationsection?view=iis-dotnet>

To obtain all the add tags for oval:mil.spawar.microsoft.iis85.site:obj:301, use:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.webServer/defaultDocument").ChildElements["files"] | Where-Object {$_.ElementTagName -eq "add"}

That covers all of the objects in ISSUE 1.  

Regarding ISSUE 2:

I cursorily observed the real path of the section using the /text:all switch, and obtained the value you were looking for in oval:mil.spawar.microsoft.iis85.site:obj:170 using:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.web/trust").GetAttribute("level").Value

I’ll investigate the remainder of issue 2 and the other issues hopefully as time permits this week.  However, I wanted to get this update out in case reading the above suggested solutions to some of the other issues you experienced using the API, which you might be able to explore faster than I can.

I suspect, however, that there is nothing appcmd.exe retrieves that is not also obtainable via the API — it’s really just a matter of figuring out how.

There is a larger question, of course, about object design.

From what I’m seeing, there are indeed two different objects that seem to be called for.  One would be an iis_site_object corresponding to Microsoft.Web.Administration.Site and its child elements, and another would be an iis_site_configuration_section_object corresponding to Microsoft.Web.Administration.ConfigurationSection and its child elements.

From there, there is a question as to how to design exploration of the element tree, or whether we really even need to support that.  We could decide that these objects require specifying concrete child entity paths.

If we design these objects well enough, it may be possible to use either the API or appcmd.exe to implement their collection.

Best regards,
—David Solin

**Comment by solind** [comment]
Hi Bryan,

I took at little time this evening to look at the rest of your issues.

Continuing ISSUE 2:

The data for oval:mil.spawar.microsoft.iis85.site:obj:322 can be obtained via:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.web/compilation").GetAttribute("Debug").Value

The data for oval:mil.spawar.microsoft.iis85.server:obj:270 can be obtained via:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.web/machineKey").GetAttribute("validation").Value

Unfortunately this is a numeric value, but the numeric values are documented here:
https://docs.microsoft.com/en-us/dotnet/api/system.web.configuration.machinekeyvalidation?redirectedfrom=MSDN&view=netframework-4.8 <https://docs.microsoft.com/en-us/dotnet/api/system.web.configuration.machinekeyvalidation?redirectedfrom=MSDN&view=netframework-4.8>

The data for oval:mil.spawar.microsoft.iis85.server:obj:271 can be obtained via:
$iis.Sites["Default Web Site"].GetWebConfiguration().GetSection("system.web/machineKey").GetAttribute("decryption").Value


Pertaining to ISSUE 3:

To get all the VDIR physical paths on the server for oval:mil.spawar.microsoft.iis85.site:obj:585, use:
$iis.Sites | %{$_.Applications} | %{$_.VirtualDirectories} | %{$_.PhysicalPath}

oval:mil.spawar.microsoft.iis85.site:obj:583 seems to rely on oval:mil.spawar.microsoft.iis85.site:obj:584, which is identical to oval:mil.spawar.microsoft.iis85.site:obj:585 above.

The structure of this retrieval suggests we may want to add an iis_virtual_directory_object to round out the other object types.


Pertaining to ISSUE 4:

The issue with the data retrieved via API for the object oval:mil.spawar.microsoft.iis85.server:obj:20 is not entirely a new one. Traditionally in OVAL we end up creating XML Enumerations for these integer flag values. This would simplify content development, but it would also mean creating special tests corresponding to each type of flag we want to make retrievable via content.  However, direct testing for discrete flags present within integer data is possible in OVAL using the “bitwise and” and “bitwise or” operators in conjunction with state entities — that is why these operators are available in the language.

So, at this point I’ve finished your option 1.  Let me know what your thoughts are, and how you think we should proceed from an object design perspective.

Best regards,
—David Solin

**Comment by bryanlwilson** [comment]
[IIS_appcmdlistconfig_test_Jan2021.zip](https://github.com/OVAL-Community/OVAL/files/5816853/IIS_appcmdlistconfig_test_Jan2021.zip)

I have finalized the appcmd test based on comments received here. The zip file is attached including example content and schema updates. Note the following updates were made to this version.

1) There were issues with some of the descriptions noting that xsi:nil="true" when the item element didn't require this. Those "xsd:documentation" field entries have been updated.
2) Updated Item Element "identifier" to allow xsi:nil="true" if "identifier_type" is set to webserver.
3) Content updated to use "result" instead of already removed "result_data" field.
4) Content updated for webserver tests where "identifier" is set to xsi:nil="true".
